SailfishOS support based on SDL2

savegame opened the pull request:
PThis PR add to Godot Engine support for exporting Godot GLES2 games to SailfishOS binary format.  
With this export templates already exists few demos and games:
- [Whoosh! in the Dungeon](https://openrepos.net/content/sashikknox/whoosh-dungeon) - Summary it's just another dungeon crawler!
- [Godot Joypad Demo](https://openrepos.net/content/sashikknox/godot-joypad) for testing gamepads support on SailfishOS (little bit modified for touchscreens)
- [Godot Materials Demo](https://openrepos.net/content/sashikknox/godot-materials-tester) - little bit modified for touchscreens

#### PR bugs fixed :
- ~~clang-fomat code styling~~ **fixed**
- ~~fix glitches on screen rotation~~ **fixed**

~~This pull request fix 26018 issue~~
I am using sources from two pull request 
 - https://github.com/godotengine/godot/pull/16470
 - https://github.com/godotengine/godot/pull/9435  

and build Godot engine for SaiflsihOS. 

#### **What is done**:
- in progress export from editor to RPM (now works only with 3.0.7 SDK release with Saiflsih-3.2.1.20-armv7hl target - becuse its hardcoded ;-) )
- touch screen input
- dynamic orientation events ( Sailfish 3.* have update SDL2 to 2.0.9 :+1: ) 
- dynamic orientation change ( with SDL_SetHint(SDL_HINT_QTWAYLAND_CONTENT_ORIENTATION,"") - not documented in SDL2 wiki, only in **SDL_Hints.h** header )
- normal application life cycle 
- sound initialization
- joypad (just compile already exists **libudev** implementation of LinuxJoypad, works nice )
 
#### **What is not done** (in next update):
- device vibrations 
- use d-bus for disable screen blank (sleep) while engine run in foreground
- ~~pause engiene when application not in foreground~~ **fixed** 
- gyroscope events (not implemented in SDL2 (not try in new SDL2 :question:  ) )
- ~~build RPM with exported project~~ (now i use [make_rpm.sh](https://github.com/savegame/sailfish_godot_rpm/blob/master/make_rpm.sh) script, and [that instructions](https://github.com/savegame/godot/releases/tag/3.2.beta-sailfish-1)  )~~  - **in progress**
- show native keyboard (maliit keyboard) when use text edit UI items 

#### **Known bugs**:
- dont work on Jolla 1, cant debug it, dont have that device 
- ~~have shader compilation error on Jolla Tablet (i try to fix it, now wait test result from users - becuse dont have that device by my self)~~ **fixed**
- SailfishOS platfrom dont support GLES3 for now, waiting for [pull request to libhybris](https://github.com/libhybris/libhybris/pull/361)

#### **Build Instructions**:

For build this branch on SailfishOS, need SailfishSDK (or Sailfish Docker image), all build commands are looks like:
```bash
export SAILFISH_SDK=$HOME/SailfishOS/
export SSH_MER_KEY=${SAILFISH_SDK}/vmshare/ssh/private_keys/engine/mersdk
export MERSSH_PORT=2222
export MERSSH="ssh  -o "IdentitiesOnly=yes" -i \"$SSH_MER_KEY\" -p $MERSSH_PORT"
# GODOT_PATH - path to godot sources inside MerSDK (SailfishSDK) build machine
export GODOT_PATH=/home/mersdk/share/Projects/Sailfish/godot
# version of taget Sailfish platform, should be more than or equal to 3.1.0.12 (becuse of SDL2 version)
export MER_TARGET_VERSION=3.2.0.12
# architecture can be i486 or armv7hl
export MER_TARGET_ARCH=armv7hl
export GODOT_ARCH=`[ "$MER_TARGET_ARCH" == "armv7hl" ] && echo arm || echo x86`
export MER_TARGET=SailfishOS-${MER_TARGET_VERSION}-${MER_TARGET_ARCH}
#first install dependencies
$MERSSH "sb2 -t ${MER_TARGET} -m sdk-install -R zypper in -y SDL2-debugsource SDL2-devel libaudioresource-devel pulseaudio-devel openssl-devel wayland-devel libpng-devel libxkbcommon-devel libudev-devel scons  gcc-c++"
#than build 
$MERSSH "cd $GODOT_PATH && sb2 -t ${MER_TARGET}  scons -j 8 platform=sailfish target=release arch=${GODOT_ARCH} bits=32"
# then optimize binary size if it release (arm or x86)
$MERSSH "cd $GODOT_PATH && sb2 -t ${MER_TARGET}  strip godot.sailfish.opt.${GODOT_ARCH}"
```
and [here](https://github.com/CODeRUS/docker-sailfishos-buildengine) is instructions how use docker with sailfish SDK (thanks to @CODeRUS :+1: )

[Latest build release](https://github.com/savegame/godot/releases/tag/3.1.1-sailfsih-2)

savegame committed:
add sdl sailfish platform

savegame committed:
add sdl sailfish platform

Commit:
- disable build GLES3 driver? when platform is sailfish

Commit:
- add libaudioresource initialization

Commit:
- add libaudioresource initialization

Commit:
- disable ALSA driver

Commit:
fix libvpx name in detect.py

Commit:
- add some verbose output

Commit:
- some fixes

Commit:
- touch events

savegame committed:
-fix build, and fix freeze when close sailfish application

savegame committed:
some fixes for landscape mode

savegame committed:
- add fixes of scene_tree for force_landscape mode

savegame committed:
+ add touch screen hint "true" to Sailfish
+ add support of landscape orientation in Sailfish (qt_extended_surface)
+ add mouse pos transform for specific orientations
- fix render buffer pixel bit size (from 16 bit to 32 bit)

savegame committed:
- full fix for forced landscape mode

savegame committed:
- fix sailfish get_window_size for landscape

savegame committed:
- force landscape options

savegame committed:
- fix viewport for normal orientations mode build

savegame committed:
- fix touch pos for portrit mode on sailfish
- fix render viewport on portrait mode on sailfish

savegame committed:
- add logo
- add fix boot image
- add fixes for touch indexes (in sfos any touch has it own unique index, not 0 - 10 )

savegame committed:
- fix build on SailfishOS SDK

savegame committed:
+ add platform libwebp-devel dependency in Sailfish SDK

savegame committed:
- add readme
- add INCLUDE search paths for different SDL2 builds

savegame committed:
- little bit fix README and detect.py

savegame committed:
+ add some missed SDL window events

savegame committed:
try build libvpx as static

savegame committed:
- disable builtin libvpx
- little bit refactoring

savegame committed:
Revert add SDL_src

savegame committed:
-remove SDL paths, no need more

savegame committed:
GLES2 shader compilation error on SailfishOS

savegame committed:
- fix built libvpx static inside Godot

savegame committed:
+ add builtin webp compilation

savegame committed:
+ build opus

savegame committed:
- litle fixed, but seems SDL2 not work more normal on SailfishOS

savegame committed:
little fix

savegame committed:
- add automatic rotation of render, when we rotate device

savegame committed:
gamepad support in SailfishOS

savegame committed:
- try fix shader compilation error on Jolla tablet

savegame committed:
- fix build, after last rebase

savegame commented:
what i need to configure travis build for SailfishOS ?  is it possible install  SailfishSDK on CI server?

akien-mga commented:
A few quick comments, I haven't reviewed much yet:

- Any PR adding new features should be done against the `master` branch. We're about to release `3.2`, so having this merged in the `3.1` branch would not make sense.
- Bundling the whole of SDL, and that only for a relatively niche platform, is a no-go. I'm sure the SailfishOS SDK provides SDL already to link against. And I assume that you might even be able to dynamically link SDL on SailfishOS so that you don't even need to distribute it, it will use the OS version.

savegame commented:
Yes, the SDL link dynamically, i don't build my own SDL version, i just include some private headers from SDL, because i need some Wayland features from SDL sources. But now, i try rewrite it without using private headers. 
OK, i will rebase it all to master branch.

savegame committed:
screen_orintetion
- successfully rewrite screen orientation changing (without SDL2 private headers)

savegame commented:
looks like i add many uneeded binary data, maybe better make squash commit from that all? orr may be i need recreate branch with "clean" commits?

CODeRUS commented:
git rebase -i is your friend ;)

savegame committed:
SailfishQt
- some work for using Qt for create Window with EGL Context

savegame committed:
build for 3.1 and higher version
- fix for build with SDL 2.0.9

savegame committed:
fixes for build

savegame committed:
seems SAILFISH gles2 dont support glRenderbufferStorageMultisample

savegame commented:
need help: 
what i need to do? for disable MXAA feature in Sailfish? 
it seems in Sailfish, libGLESv2 dont have glRenderbufferStorageMultisample ( i try `nm -D libGLESv2.so|grep glRenderbufferStorageMultisample` )
now its compile well and work, but i have glitches, and error:
```
WARNING: _render_target_allocate: Cannot allocate mipmaps for 3D post processing effects
   At: drivers/gles2/rasterizer_storage_gles2.cpp:4916.
```

savegame committed:
remove unused

savegame committed:
reset wrong changes & add SAILFISH_ENABLED define

savegame committed:
- fix merge error ( deleted some test code )

savegame commented on scene/main/scene_tree.cpp:
```diff
@@ -1232,6 +1278,9 @@ void SceneTree::_update_root_rect() {
 
+			// force landscape 
+			screen_size = Size2( screen_size.y, screen_size.x );
+			root->set_attach_to_screen_rect(Rect2(margin, screen_size));
```
merge bug

savegame committed:
- disable some fetures for Sailfish
- fix libaudioresource initialization text outpeut

savegame commented:
hi everyone... 
i have troubles with master branch ) 
i merge and build Saiflish port from 3.1 branch to mater. On 3.1 my code works fine, but on master i have glitches:

| Sailfish Godot 3.1 | Sailfish Godot 3.2 |
| --- | --- |
| [ ![](https://i.ibb.co/YpfHmdJ/2019-11-26-11-58.png) ](https://ibb.co/YpfHmdJ)  | [![](https://i.ibb.co/4V4ZF9T/20191125-001.png)](https://ibb.co/4V4ZF9T) |
| [ ![](https://i.ibb.co/ZN68YPf/20191126-001.png) ](https://ibb.co/ZN68YPf) | [ ![](https://i.ibb.co/L6s71NH/20191124-001.png) ](https://ibb.co/L6s71NH) |

as i know, reason is my - hack for SailfishOS landscape orientation, but seems now rotate only part of canvas on scene (from shader i mean), other items draw in default screen orientation  ... how can i understand in code, where i should draw finished scene buffer, and when i draw any canvas item (which should by rotated too, as all scene)
p.s. as i understand, in 3.1 all items was first drawn to scene buffer, then scene buffer was draw to screen, but in 3.2 items draw to screen directly? 
all my code with "hack" inside definitions **SAILFSIH_FORCE_LANDSCAPE**, and **gles2/shaders/canvas.glsl** 
@akien-mga, if you can, please tell where i should look, to fix my bug with drawing scene in landscape mode... 
3d now looks not works

savegame commented:
now i see, i modified canvas.glsl shader, and every item on canvas draw with this shader... now i just need understand where in gles2_rasterizer draw final canvas buffer, and draw it with landscape UV modifications, and other canvas items draw normally.

savegame commented:
:tada:  i found an answer! ) 
just need reset FORCE_LANDSCAPE in beggining of ""_canvas_item_render_commands"" function ?and all draw correctly now!

savegame committed:
- fix glitches in SailfishOS Landscape mode

savegame committed:
+ clang-format

savegame commented:
waiting for desigion  of maintainers :nerd_face:

savegame commented:
@akien-mga,  is it need create new **PR** for this (i'm force rewrite my **3.1_sailfish_sdl2** branch with **master** branch), or just live it here as is?

akien-mga commented:
It seems ok, the branch name doesn't matter much. What's important is the target branch, which is properly tracking `master`.

savegame committed:
remove test option for lock transpose function in GLES2 canvas shader, for Salfish i486

savegame committed:
- fix Jolla Tablet shader compilation
 On Jolla Tablet x86 transpose fucntion have conflict name with builtin transpose function

savegame commented:
can any one restart travis check, dont know why mono is not build

savegame committed:
in progress creating export project from editor to Sailfish rpm file

savegame committed:
in progress - create exporter to SailfishOS rpm's

savegame committed:
in progress - Sailfish export
  - add list mer targets, and verifying

savegame committed:
in progress - export to Sailfish

savegame committed:
in progress Sailfish Export

savegame committed:
in progress Sailfish export
- good, now we  have right configured environment for build RPM fish sfdk

savegame committed:
first release of Sailfish export from editor

savegame committed:
- add option, for choose assets install prefix

savegame committed:
in progress SailfishOS export from Editor

- add export to both saiflish platforms (arm, x86)

savegame committed:
- generate package name
- check package name by name validator [a-z_-0-9\.]+

akien-mga commented:
Superseded by #38642.
